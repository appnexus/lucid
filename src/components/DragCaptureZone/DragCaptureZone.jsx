import _ from 'lodash';
import React from 'react';
import { lucidClassNames } from '../../util/style-helpers';
import { createClass } from '../../util/component-definition';

const boundClassNames = lucidClassNames.bind('&-DragCaptureZone');
const {
	func,
	string
} = React.PropTypes;

/**
 * {"categories": ["utility"]}
 *
 * This is a helper component used to capture mouse events to determine when the
 * user starts, is and stops dragging.
 */
const DragCaptureZone = createClass({
	displayName: 'DragCaptureZone',
	propTypes: {
		/**
		 * Appended to the component-specific class names set on the root
		 * element.
		 */
		className: string,

		/**
		 * Called as the user drags the mouse.
		 *
		 * Signature: `({ dx, dy, pageX, pageY }, { event, props }) => {}`
		 */
		onDrag: func,

		/**
		 * Called when the user releases the mouse button after having dragged.
		 *
		 * Signature: `({ dx, dy, pageX, pageY }, { event, props }) => {}`
		 */
		onDragEnd: func,

		/**
		 * Called when the user presses the mouse button down while over the
		 * component.
		 *
		 * Signature: `({ dx, dy, pageX, pageY }, { event, props }) => {}`
		 */
		onDragStart: func
	},

	getDefaultProps() {
		return {
			onDrag: _.noop,
			onDragEnd: _.noop,
			onDragStart: _.noop
		};
	},

	getInitialState() {
		return {
			pageX: 0,
			pageY: 0
		};
	},

	render() {
		return (
			<div
				{...this.props}
				className={boundClassNames('&', this.props.className)}
				key='DragCaptureZone'
				onMouseDown={this.handleDragStart}
			/>
		);
	},

	componentWillUnmount() {
		window.document.removeEventListener('mousemove', this.handleDrag);
		window.document.removeEventListener('mouseup', this.handleDragEnd);
	},

	handleDrag(event) {
		const {
			pageX,
			pageY
		} = event;

		event.preventDefault();

		this.props.onDrag({
			dX: pageX - this.state.pageX,
			dY: pageY - this.state.pageY,
			pageX,
			pageY
		}, {
			event,
			props: this.props
		});
	},

	handleDragEnd(event) {
		const {
			pageX,
			pageY
		} = event;

		event.preventDefault();

		window.document.removeEventListener('mousemove', this.handleDrag);
		window.document.removeEventListener('mouseup', this.handleDragEnd);

		this.props.onDragEnd({
			dX: pageX - this.state.pageX,
			dY: pageY - this.state.pageY,
			pageX,
			pageY
		}, {
			event,
			props: this.props
		});

		this.setState({
			pageX: 0,
			pageY: 0
		});
	},

	handleDragStart(event) {
		const {
			pageX,
			pageY
		} = event;

		event.preventDefault();

		window.document.addEventListener('mousemove', this.handleDrag);
		window.document.addEventListener('mouseup', this.handleDragEnd);

		this.props.onDragStart({
			dX: 0,
			dY: 0,
			pageX,
			pageY
		}, {
			event,
			props: this.props
		});

		this.setState({
			pageX,
			pageY
		});
	}
});

export default DragCaptureZone;
