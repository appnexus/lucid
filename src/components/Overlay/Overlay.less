@import (reference) '../../styles/variables.less';
@import (reference) '../../styles/mixins.less';

.lucid-Overlay {
	.transition-group-animation-fade();
	display: flex;
	justify-content: center;
	align-items: center;
	position: fixed;
	left: 0;
	top: 0;
	width: 100vw;
	height: 100vh;
	overflow: hidden;
	z-index: @zindex-modal;
	background-color: @color-gray-60;

	&&-is-not-modal {
		background-color: @color-transparent;
		// Setting this to `none` allows us to use flex box to center the content
		// on the screen while still allowing the user to interact with elements
		// behind it
		pointer-events: none;

		// Have to set the pointer events back to initial so the user can interact
		// with the contents
		> * {
			pointer-events: auto;
		}
	}
}
