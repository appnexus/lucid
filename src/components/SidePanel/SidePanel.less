@import (reference) '../../styles/variables.less';
@import (reference) '../../styles/mixins.less';

.@{prefix}-SidePanel {

	font-family: @fontFamily;

	.@{prefix}-SidePanel-pane {
		position: fixed;
		top: 0;
		display: flex;
		flex-direction: column;
		height: 100%;
		width: 240px;
		background-color: @color-neutral-2;
		box-shadow: -1px 0px 5px 0px rgba(33, 31, 31, 0.3);

		.@{prefix}-SidePanel-is-animated& {
			transition: transform 200ms;
		}

		.@{prefix}-SidePanel-is-expanded.@{prefix}-SidePanel-is-expanded& {
			transform: translateX(0);
		}

		.@{prefix}-SidePanel-position-right& {
			right: 0;
			transform: translateX(100%);
		}

		.@{prefix}-SidePanel-position-left& {
			left: 0;
			flex-direction: row-reverse;
			transform: translateX(-100%);
		}
		
		.@{prefix}-SidePanel-header {
			border-bottom: @border-lightBorder;
			background: white;
			padding: 20px 20px 0 20px;
			
			.@{prefix}-SidePanel-header-inner-wrapper {
				display: flex;
				
				.@{prefix}-SidePanel-header-content {
					flex: 1;
					height: 100%;
				}

				.@{prefix}-Button {
					padding: 3px;
				}
				
				.@{prefix}-SidePanel-header-closer {
					fill: @color-disabledText;

					&.@{prefix}-CrossIcon {
						// fill: @color-disabledText;
						margin: 0;
					}
				}
			}
		}
		.@{prefix}-SidePanel-body {

			height: 100%;
			display: flex;

			.@{prefix}-SidePanel-grabber {
				background-color: @color-neutral-2;
				width: 20px;
				height: 100%;
				display: flex;
				justify-content: center;
				align-items: center;
				cursor: grab;
				box-sizing: border-box;
	
				.@{prefix}-GripperVerticalIcon {
					transform: translateX(-0.5px);
					opacity: 0.5;
				}
	
				&:active {
					cursor: grabbing;
				}
	
				&:active, &:hover {

					.@{prefix}-GripperVerticalIcon {
						opacity: 1;
					}
				}
			}

			.@{prefix}-SidePanel-content {
				overflow: auto;
				padding: 20px 20px 20px 0;
				flex: 1;
			}

		}
		
	}
}
