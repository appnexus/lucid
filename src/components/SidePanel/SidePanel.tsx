import _, { omit } from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';

import { lucidClassNames } from '../../util/style-helpers';
import Overlay, { IOverlayProps } from '../Overlay/Overlay';
import GripperVerticalIcon from '../Icon/GripperVerticalIcon/GripperVerticalIcon';
import CloseIcon from '../Icon/CloseIcon/CloseIcon';
import DragCaptureZone from '../DragCaptureZone/DragCaptureZone';
import Button from '../Button/Button';
import { getFirst, StandardProps } from '../../util/component-types';

const cx = lucidClassNames.bind('&-SidePanel');

const { any, bool, func, oneOf, node, number, string, oneOfType } = PropTypes;

/* SidePanel Header **/
const SidePanelHeader = (_props: StandardProps): null => null;
SidePanelHeader.displayName = 'SidePanel.Header';
SidePanelHeader.propName = 'Header';
SidePanelHeader.peek = {
	description: `
		Defines the Header content of SidePanel. If no content is defined, it will
		still show the close button.
	`,
};
SidePanelHeader.propTypes = {
	/**
		Children that will be rendered.
	*/
	children: node,
};

/** Side Panel */
export interface ISidePanelProps extends Partial<IOverlayProps> {
	/** Alternative to using `<SidePanel.Header>`. */
	Header?: string | (React.ReactNode & { props: StandardProps });

	/** Controls the expanded/collapsed state as a boolean prop. */
	isExpanded: boolean;

	/** When true, hides the resizer at the edge of the SidePanel. */
	isResizeDisabled: boolean;

	/** Callback triggered when user clicks the background, hits the Esc key, or
	 * clicks the close button in the Header. */
	onCollapse: ({
		event,
		props,
	}: {
		event: React.MouseEvent | KeyboardEvent;
		props: ISidePanelProps;
	}) => void;

	/** Callback triggered after a user resizes to a new width. */
	onResize: (
		width: number,
		{
			event,
			props,
		}: {
			event: MouseEvent | TouchEvent;
			props: ISidePanelProps;
		}
	) => void;

	/** Controls the position on the screen. */
	position: 'left' | 'right';

	/** When true, it will prevent scrolling in the background when \`isExpanded\;
	 * is true. This is accomplished by setting \`document.body.style.overflow =
	 * 'hidden'\`. */
	preventBodyScroll: boolean;

	/** Sets the initial width in pixels. The actual width may change if the user
	 * resizes it. */
	width: number;

	/** Sets the minimum width in pixels. */
	minWidth: number;

	/** Sets the maximim width in pixels. */
	maxWidth: number;

	/** Sets the top margin for the panel. Defaults to \`0\`. */
	topOffset: number | string;
}

/** TODO: Remove the 'nonPassThroughs' when the component is converted to a functional component */
const nonPassThroughs = [
	'children',
	'className',
	'Header',
	'isAnimated',
	'isExpanded',
	'isResizeDisabled',
	'onCollapse',
	'onResize',
	'position',
	'preventBodyScroll',
	'width',
	'minWidth',
	'maxWidth',
	'topOffset',
	'initialState',
];

interface ISidePanelState {
	isResizing: boolean;
	width: number;
	startWidth: number;
	isExpanded: boolean;
}

class SidePanel extends React.Component<ISidePanelProps, ISidePanelState, {}> {
	static displayName = 'SidePanel';
	static peek = {
		description: `A fixed-positioned overlay positioned on the side of the screen at full screen height. Supports variable widths resized by the user or defined as a prop. Animated collapse and expand with optional Header and closer.`,
		categories: ['layout'],
	};
	static propTypes = {
		/**
			Content of the SidePanel, but also accepts \`<SidePanel.Header>\` to define
			header content.
		*/
		children: node,

		/**
			Appended to the component-specific class names set on the root element.
		*/
		className: string,

		/**
			Alternative to using \`<SidePanel.Header>\`.
		*/
		Header: any,

		/**
			Enables animated transitions during expansion and collapse.
		*/
		isAnimated: bool,

		/**
			Controls the expanded/collapsed state as a boolean prop.
		*/
		isExpanded: bool,

		/**
			When true, hides the resizer at the edge of the SidePanel.
		*/
		isResizeDisabled: bool,

		/**
			Callback triggered when user clicks the background, hits the Esc key, or
			clicks the close button in the Header.
			Signature: \`({ event, props }) => {}\`
		*/
		onCollapse: func,

		/**
			Callback triggered after a user resizes to a new width.
			Signature: \`(width, { event, props }) => {}\`
		*/
		onResize: func,

		/**
			Controls the position on the screen.
		*/
		position: oneOf(['left', 'right']),

		/**
			When true, it will prevent scrolling in the background when \`isExpanded\`
			is true. This is accomplished by setting \`document.body.style.overflow =
			'hidden'\`.
		*/
		preventBodyScroll: bool,

		/**
			Sets the initial width in pixels. The actual width may change if the user
			resizes it.
		*/
		width: number,

		/**
			Sets the minimum width of the Side Panel.
		*/
		minWidth: number,

		/**
			Sets the maximum width of the Side Panel.
		*/
		maxWidth: number,

		/**
			Sets the top margin for the panel. Defaults to \`0\`.
		*/
		topOffset: oneOfType([number, string]),
	};

	state = {
		isResizing: false,
		width: this.props.width as number,
		startWidth: this.props.width as number,
		isExpanded: this.props.isExpanded as boolean,
	};

	static defaultProps = {
		isAnimated: true,
		isExpanded: true,
		isResizeDisabled: false,
		onCollapse: _.noop,
		onResize: _.noop,
		position: 'right',
		preventBodyScroll: false,
		topOffset: 0,
		width: 240,
		minWidth: -Infinity,
		maxWidth: Infinity,
	};

	static Header = SidePanelHeader;

	timerId = setTimeout((): void => {
		return;
	}, 1);

	handleResizeStart = (): void => {
		this.setState({
			isResizing: true,
		});
	};

	handleResize = ({ dX }: { dX: number }): void => {
		const { startWidth } = this.state;
		const position =
			startWidth + dX * (this.props.position === 'right' ? -1 : 1);
		this.setState({
			width: _.clamp(
				position,
				this.props.minWidth,
				this.props.maxWidth === Infinity
					? window.innerWidth
					: this.props.maxWidth
			),
		});
	};

	handleResizeEnd = (
		{ dX }: { dX: number },
		{ event }: { event: MouseEvent | TouchEvent }
	): void => {
		const { startWidth, width } = this.state;
		this.setState({
			width: startWidth + dX * (this.props.position === 'right' ? -1 : 1),
			isResizing: false,
			startWidth: width,
		});
		this.props.onResize(startWidth - dX, { props: this.props, event });
	};

	handleCollapse = ({
		event,
	}: {
		event: React.MouseEvent | KeyboardEvent;
	}): void => {
		this.props.onCollapse({ event, props: this.props });
	};

	componentDidUpdate(prevProps: ISidePanelProps): void {
		if (prevProps.isExpanded !== this.props.isExpanded) {
			this.timerId = setTimeout((): void => {
				this.setState({
					isExpanded: this.props.isExpanded as boolean,
				});
			}, 1);
		}
	}

	componentWillUnmount(): void {
		const { preventBodyScroll } = this.props;
		if (this.timerId) {
			clearTimeout(this.timerId);
		}
		if (preventBodyScroll) {
			window.document.body.style.overflow = '';
		}
	}

	render(): React.ReactNode {
		const {
			children,
			className,
			isAnimated,
			isExpanded,
			isResizeDisabled,
			position,
			preventBodyScroll,
			topOffset,
			...passThroughs
		} = this.props;

		const headerEl = getFirst(this.props, SidePanel.Header);
		const headerChildren = _.get(headerEl, 'props.children');

		if (preventBodyScroll) {
			window.document.body.style.overflow = isExpanded ? 'hidden' : '';
		}

		return (
			<Overlay
				className={cx(
					'&',
					{
						'&-is-expanded': isExpanded && this.state.isExpanded,
						'&-position-left': position === 'left',
						'&-position-right': position === 'right',
						'&-is-animated': isAnimated,
					},
					className
				)}
				isShown={isExpanded || this.state.isExpanded}
				onBackgroundClick={this.handleCollapse}
				onEscape={this.handleCollapse}
				isAnimated={isAnimated}
				style={{
					marginTop: topOffset,
				}}
				{...omit(passThroughs, nonPassThroughs)}
			>
				<div
					className={cx('&-pane')}
					style={{
						width: this.state.width,
						marginTop: topOffset,
					}}
				>
					{headerEl && (
						<div className={cx('&-header')}>
							<div className={cx('&-header-inner-wrapper')}>
								<div className={cx('&-header-content')}>{headerChildren}</div>

								<Button
									className={cx('&-header-closer-button')}
									kind='invisible'
									onClick={this.handleCollapse}
									hasOnlyIcon={true}
								>
									<CloseIcon
										className={cx('&-header-closer')}
										isClickable
										size={14}
									/>
								</Button>
							</div>
						</div>
					)}

					<div className={cx('&-body')}>
						{!isResizeDisabled && (
							<DragCaptureZone
								className={cx('&-grabber')}
								onDragStart={this.handleResizeStart}
								onDrag={this.handleResize}
								onDragEnd={this.handleResizeEnd}
							>
								<GripperVerticalIcon width='20' />
							</DragCaptureZone>
						)}
						<div className={cx('&-content')}>{children}</div>
					</div>
				</div>
			</Overlay>
		);
	}
}

export default SidePanel;
