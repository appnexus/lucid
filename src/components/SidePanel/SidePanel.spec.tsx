import _, { forEach, has } from 'lodash';
import React from 'react';
import { mount, shallow } from 'enzyme';

import { common } from '../../util/generic-tests';
import SidePanel from './SidePanel';
import Overlay from '../Overlay/Overlay';
import DragCaptureZone from '../DragCaptureZone/DragCaptureZone';

describe('SidePanel', () => {
	common(SidePanel);

	describe('pass throughs', () => {
		let wrapper: any;
		const defaultProps = SidePanel.defaultProps;

		beforeEach(() => {
			const props = {
				...defaultProps,
				className: 'wut',
				isModal: false,
				isExpanded: false,
				isAnimated: false,
				isResizeDisabled: true,
				position: 'left' as any,
				preventBodyScroll: true,
				topOffset: 10,
				width: 1000,
				minWidth: 500,
				maxWidth: 2000,
				style: { marginRight: 10 },
				initialState: { test: true },
				callbackId: 1,
				'data-testid': 10,
				Header: (
					<div>
						<strong>Rich content</strong>
					</div>
				),
			};
			wrapper = shallow(<SidePanel {...props} />);
		});

		afterEach(() => {
			wrapper.unmount();
		});

		it('passes through props not defined in `propTypes` to the root element.', () => {
			const rootProps = wrapper.find('.lucid-SidePanel').props();

			// Note: 'className', 'style' and 'isAnimated' are plucked from the pass through object
			// but still appear becuase they are also directly passed to the root `Overlay` element as a prop.
			// 'isShown' appears because it is calculated based on the 'isExpanded' prop.
			// The root `Overlay` element is not a DOM element so 'callbackId', if it exists, is also passed through.
			forEach(
				[
					'className',
					'isShown',
					'onEscape',
					'isAnimated',
					'style',
					'isModal',
					'data-testid',
					'callbackId',
					'children',
				],
				(prop) => {
					expect(has(rootProps, prop)).toBe(true);
				}
			);

			expect(wrapper.first().prop(['className'])).toContain('wut');
			expect(wrapper.first().prop(['isShown'])).toBe(false); // set based on isExpaneded
			expect(wrapper.first().prop(['isAnimated'])).toBe(false);
			expect(wrapper.first().prop(['isModal'])).toBe(false);
			expect(wrapper.first().prop(['style'])).toMatchObject({
				marginRight: 10,
			});
			expect(wrapper.first().prop(['data-testid'])).toBe(10);
			expect(wrapper.first().prop(['callbackId'])).toBe(1);
		});

		it('omits the props defined in `propTypes` from the root element, plus, in addition, `initialState`.', () => {
			const rootProps = wrapper.find('.lucid-SidePanel').props();

			forEach(
				[
					'Header',
					'isExpanded',
					'isResizeDisabled',
					'onCollapse',
					'onResize',
					'position',
					'preventBodyScroll',
					'width',
					'minWidth',
					'maxWidth',
					'topOffset',
					'initialState',
				],
				(prop) => {
					expect(has(rootProps, prop)).toBe(false);
				}
			);
		});
	});

	describe('Events', () => {
		describe('onCollapse', () => {
			it('should be called when Overlay background is clicked', () => {
				const onCollapse = jest.fn();
				const wrapper = shallow(<SidePanel onCollapse={onCollapse} />);
				const overlayWrapper = wrapper.find(Overlay);
				const mockEvent: any = new Event('click');
				overlayWrapper.prop('onBackgroundClick')({
					event: mockEvent,
					props: overlayWrapper.props(),
				});
				expect(onCollapse).toHaveBeenCalled();
			});

			it('should be called when Escape key is pressed', () => {
				const onCollapse = jest.fn();
				const wrapper = shallow(<SidePanel onCollapse={onCollapse} />);
				const overlayWrapper = wrapper.find(Overlay);
				const mockEvent: any = new Event('keypress');
				overlayWrapper.prop('onEscape')({
					event: mockEvent,
					props: overlayWrapper.props(),
				});
				expect(onCollapse).toHaveBeenCalled();
			});

			it('should be called when the close icon is clicked', () => {
				const onCollapse = jest.fn();
				const wrapper = shallow(
					<SidePanel isExpanded onCollapse={onCollapse} Header='Foo bar' />
				);
				const crossIconWrapper = wrapper.find(
					'.lucid-SidePanel-header-closer-button'
				);
				crossIconWrapper.simulate('click', { event: 'abc' });
				expect(onCollapse).toHaveBeenCalled();
			});
		});

		describe('onResize', () => {
			it('should be called when the resizer stops being dragged', () => {
				const onResize = jest.fn();
				const wrapper = shallow(<SidePanel onResize={onResize} />);
				const dragCaptureZoneWrapper = wrapper.find(DragCaptureZone);
				const mockEvent = new Event('mouseup');
				dragCaptureZoneWrapper.prop('onDragEnd')(
					{
						dx: 100,
						dy: 0,
						pageX: 100,
						pageY: 0,
					} as any,
					{
						event: mockEvent,
						props: dragCaptureZoneWrapper.props(),
					} as any
				);
				expect(onResize).toHaveBeenCalled();
			});
		});
	});

	describe('preventBodyScroll', () => {
		it('should hide the body overflow to prevent scrolling', () => {
			mount(<SidePanel isExpanded preventBodyScroll />);
			expect(document.body.style.overflow).toEqual('hidden');
		});

		it('should hide the body overflow to prevent scrolling', () => {
			const wrapper = mount(<SidePanel isExpanded preventBodyScroll />);
			wrapper.unmount();
			expect(document.body.style.overflow).toEqual('');
		});
	});

	describe('position', () => {
		it('should match snapshot and apply the appropriate class', () => {
			const wrapper = shallow(<SidePanel position='left' />);
			expect(
				wrapper
					.find('.lucid-SidePanel')
					.hasClass('lucid-SidePanel-position-left')
			).toEqual(true);
			expect(wrapper).toMatchSnapshot();
		});

		it('should match snapshot and apply the appropriate class', () => {
			const wrapper = shallow(<SidePanel position='right' />);
			expect(
				wrapper
					.find('.lucid-SidePanel')
					.hasClass('lucid-SidePanel-position-right')
			).toEqual(true);
			expect(wrapper).toMatchSnapshot();
		});
	});

	describe('isResizeDisabled', () => {
		it('should match snapshot', () => {
			const wrapper = shallow(<SidePanel isResizeDisabled />);
			expect(wrapper).toMatchSnapshot();
		});
	});

	describe('isAnimated', () => {
		it('should match snapshot and apply the appropriate class', () => {
			const wrapper = shallow(<SidePanel isAnimated={true} />);
			expect(
				wrapper.find('.lucid-SidePanel').hasClass('lucid-SidePanel-is-animated')
			).toEqual(true);
			expect(wrapper).toMatchSnapshot();
		});
	});

	describe('isExpanded', () => {
		it('should match snapshot and apply the appropriate class', () => {
			const wrapper = shallow(<SidePanel isExpanded={true} />);

			wrapper.setState({ isExpanded: true });

			expect(wrapper).toMatchSnapshot();
			expect(
				wrapper.find('.lucid-SidePanel').hasClass('lucid-SidePanel-is-expanded')
			).toEqual(true);
		});
	});

	describe('width', () => {
		it('should match snapshot', () => {
			const wrapper = shallow(<SidePanel width={200} />);
			expect(wrapper).toMatchSnapshot();
		});
	});

	describe('Header', () => {
		it('should match snapshot', () => {
			const wrapper = shallow(<SidePanel Header='This is a header' />);
			expect(wrapper).toMatchSnapshot();
		});
	});
});
