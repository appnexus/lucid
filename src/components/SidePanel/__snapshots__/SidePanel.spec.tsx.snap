// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidePanel [common] example testing should match snapshot(s) for NoModal 1`] = `
<section>
  <p>
    modal can be disabled in the underlying Overlay component to allow interaction with the background
  </p>
  <button>
    Toggle SidePanel
  </button>
</section>
`;

exports[`SidePanel [common] example testing should match snapshot(s) for WithDefaults 1`] = `
<section>
  <p>
    expand and collapse SidePanel by passing the isExpanded prop
  </p>
  <button>
    Toggle SidePanel
  </button>
</section>
`;

exports[`SidePanel [common] example testing should match snapshot(s) for WithHeader 1`] = `
<section>
  <p>
    render a header with a close button
  </p>
  <button>
    Toggle SidePanel
  </button>
</section>
`;

exports[`SidePanel [common] example testing should match snapshot(s) for WithLeftPosition 1`] = `
<section>
  <p>
    position of the SidePanel can be aligned on either side of the page
  </p>
  <button>
    Toggle SidePanel
  </button>
</section>
`;

exports[`SidePanel [common] example testing should match snapshot(s) for WithResize 1`] = `
<section>
  <p>
    resized width is obtatined with a callback
  </p>
  <p>
    SidePanel width: 240
  </p>
  <button>
    Toggle SidePanel
  </button>
</section>
`;

exports[`SidePanel [common] example testing should match snapshot(s) for WithResponsiveGrid 1`] = `
<section>
  <p>
    modal can be disabled in the underlying Overlay component to allow interaction with the background
  </p>
  <button>
    Toggle SidePanel
  </button>
</section>
`;

exports[`SidePanel [common] example testing should match snapshot(s) for WithTopMargin 1`] = `
<section>
  <p>
    render the component with a top margin
  </p>
  <button>
    Toggle SidePanel
  </button>
</section>
`;

exports[`SidePanel [common] example testing should match snapshot(s) for WithoutAnimation 1`] = `
<section>
  <p>
    transition animations can be disabled if desired
  </p>
  <button>
    Toggle SidePanel
  </button>
</section>
`;

exports[`SidePanel [common] example testing should match snapshot(s) for WithoutResize 1`] = `
<section>
  <p>
    prevent resizing width by hiding the resizer
  </p>
  <button>
    Toggle SidePanel
  </button>
</section>
`;

exports[`SidePanel Header should match snapshot 1`] = `
<Overlay
  className="lucid-SidePanel lucid-SidePanel-is-expanded lucid-SidePanel-position-right lucid-SidePanel-is-animated"
  isAnimated={true}
  isModal={true}
  isShown={true}
  onBackgroundClick={[Function]}
  onEscape={[Function]}
  style={
    Object {
      "marginTop": 0,
    }
  }
>
  <div
    className="lucid-SidePanel-pane"
    style={
      Object {
        "marginTop": 0,
        "width": 240,
      }
    }
  >
    <div
      className="lucid-SidePanel-header"
    >
      <div
        className="lucid-SidePanel-header-inner-wrapper"
      >
        <div
          className="lucid-SidePanel-header-content"
        >
          This is a header
        </div>
        <Button
          className="lucid-SidePanel-header-closer-button"
          hasOnlyIcon={true}
          isActive={false}
          isDisabled={false}
          kind="invisible"
          onClick={[Function]}
          type="button"
        >
          <CloseIcon
            aspectRatio="xMidYMid meet"
            className="lucid-SidePanel-header-closer"
            color="primary"
            isClickable={true}
            isDisabled={false}
            onClick={[Function]}
            onSelect={[Function]}
            size={14}
            viewBox="0 0 16 16"
          />
        </Button>
      </div>
    </div>
    <div
      className="lucid-SidePanel-body"
    >
      <DragCaptureZone
        className="lucid-SidePanel-grabber"
        onDrag={[Function]}
        onDragCancel={[Function]}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <GripperVerticalIcon
          aspectRatio="xMidYMid meet"
          color="primary"
          isClickable={false}
          isDisabled={false}
          onClick={[Function]}
          onSelect={[Function]}
          size={16}
          viewBox="0 0 16 16"
          width="20"
        />
      </DragCaptureZone>
      <div
        className="lucid-SidePanel-content"
      />
    </div>
  </div>
</Overlay>
`;

exports[`SidePanel isAnimated should match snapshot and apply the appropriate class 1`] = `
<Overlay
  className="lucid-SidePanel lucid-SidePanel-is-expanded lucid-SidePanel-position-right lucid-SidePanel-is-animated"
  isAnimated={true}
  isModal={true}
  isShown={true}
  onBackgroundClick={[Function]}
  onEscape={[Function]}
  style={
    Object {
      "marginTop": 0,
    }
  }
>
  <div
    className="lucid-SidePanel-pane"
    style={
      Object {
        "marginTop": 0,
        "width": 240,
      }
    }
  >
    <div
      className="lucid-SidePanel-body"
    >
      <DragCaptureZone
        className="lucid-SidePanel-grabber"
        onDrag={[Function]}
        onDragCancel={[Function]}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <GripperVerticalIcon
          aspectRatio="xMidYMid meet"
          color="primary"
          isClickable={false}
          isDisabled={false}
          onClick={[Function]}
          onSelect={[Function]}
          size={16}
          viewBox="0 0 16 16"
          width="20"
        />
      </DragCaptureZone>
      <div
        className="lucid-SidePanel-content"
      />
    </div>
  </div>
</Overlay>
`;

exports[`SidePanel isExpanded should match snapshot and apply the appropriate class 1`] = `
<Overlay
  className="lucid-SidePanel lucid-SidePanel-is-expanded lucid-SidePanel-position-right lucid-SidePanel-is-animated"
  isAnimated={true}
  isModal={true}
  isShown={true}
  onBackgroundClick={[Function]}
  onEscape={[Function]}
  style={
    Object {
      "marginTop": 0,
    }
  }
>
  <div
    className="lucid-SidePanel-pane"
    style={
      Object {
        "marginTop": 0,
        "width": 240,
      }
    }
  >
    <div
      className="lucid-SidePanel-body"
    >
      <DragCaptureZone
        className="lucid-SidePanel-grabber"
        onDrag={[Function]}
        onDragCancel={[Function]}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <GripperVerticalIcon
          aspectRatio="xMidYMid meet"
          color="primary"
          isClickable={false}
          isDisabled={false}
          onClick={[Function]}
          onSelect={[Function]}
          size={16}
          viewBox="0 0 16 16"
          width="20"
        />
      </DragCaptureZone>
      <div
        className="lucid-SidePanel-content"
      />
    </div>
  </div>
</Overlay>
`;

exports[`SidePanel isResizeDisabled should match snapshot 1`] = `
<Overlay
  className="lucid-SidePanel lucid-SidePanel-is-expanded lucid-SidePanel-position-right lucid-SidePanel-is-animated"
  isAnimated={true}
  isModal={true}
  isShown={true}
  onBackgroundClick={[Function]}
  onEscape={[Function]}
  style={
    Object {
      "marginTop": 0,
    }
  }
>
  <div
    className="lucid-SidePanel-pane"
    style={
      Object {
        "marginTop": 0,
        "width": 240,
      }
    }
  >
    <div
      className="lucid-SidePanel-body"
    >
      <div
        className="lucid-SidePanel-content"
      />
    </div>
  </div>
</Overlay>
`;

exports[`SidePanel position should match snapshot and apply the appropriate class 1`] = `
<Overlay
  className="lucid-SidePanel lucid-SidePanel-is-expanded lucid-SidePanel-position-left lucid-SidePanel-is-animated"
  isAnimated={true}
  isModal={true}
  isShown={true}
  onBackgroundClick={[Function]}
  onEscape={[Function]}
  style={
    Object {
      "marginTop": 0,
    }
  }
>
  <div
    className="lucid-SidePanel-pane"
    style={
      Object {
        "marginTop": 0,
        "width": 240,
      }
    }
  >
    <div
      className="lucid-SidePanel-body"
    >
      <DragCaptureZone
        className="lucid-SidePanel-grabber"
        onDrag={[Function]}
        onDragCancel={[Function]}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <GripperVerticalIcon
          aspectRatio="xMidYMid meet"
          color="primary"
          isClickable={false}
          isDisabled={false}
          onClick={[Function]}
          onSelect={[Function]}
          size={16}
          viewBox="0 0 16 16"
          width="20"
        />
      </DragCaptureZone>
      <div
        className="lucid-SidePanel-content"
      />
    </div>
  </div>
</Overlay>
`;

exports[`SidePanel position should match snapshot and apply the appropriate class 2`] = `
<Overlay
  className="lucid-SidePanel lucid-SidePanel-is-expanded lucid-SidePanel-position-right lucid-SidePanel-is-animated"
  isAnimated={true}
  isModal={true}
  isShown={true}
  onBackgroundClick={[Function]}
  onEscape={[Function]}
  style={
    Object {
      "marginTop": 0,
    }
  }
>
  <div
    className="lucid-SidePanel-pane"
    style={
      Object {
        "marginTop": 0,
        "width": 240,
      }
    }
  >
    <div
      className="lucid-SidePanel-body"
    >
      <DragCaptureZone
        className="lucid-SidePanel-grabber"
        onDrag={[Function]}
        onDragCancel={[Function]}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <GripperVerticalIcon
          aspectRatio="xMidYMid meet"
          color="primary"
          isClickable={false}
          isDisabled={false}
          onClick={[Function]}
          onSelect={[Function]}
          size={16}
          viewBox="0 0 16 16"
          width="20"
        />
      </DragCaptureZone>
      <div
        className="lucid-SidePanel-content"
      />
    </div>
  </div>
</Overlay>
`;

exports[`SidePanel width should match snapshot 1`] = `
<Overlay
  className="lucid-SidePanel lucid-SidePanel-is-expanded lucid-SidePanel-position-right lucid-SidePanel-is-animated"
  isAnimated={true}
  isModal={true}
  isShown={true}
  onBackgroundClick={[Function]}
  onEscape={[Function]}
  style={
    Object {
      "marginTop": 0,
    }
  }
>
  <div
    className="lucid-SidePanel-pane"
    style={
      Object {
        "marginTop": 0,
        "width": 200,
      }
    }
  >
    <div
      className="lucid-SidePanel-body"
    >
      <DragCaptureZone
        className="lucid-SidePanel-grabber"
        onDrag={[Function]}
        onDragCancel={[Function]}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <GripperVerticalIcon
          aspectRatio="xMidYMid meet"
          color="primary"
          isClickable={false}
          isDisabled={false}
          onClick={[Function]}
          onSelect={[Function]}
          size={16}
          viewBox="0 0 16 16"
          width="20"
        />
      </DragCaptureZone>
      <div
        className="lucid-SidePanel-content"
      />
    </div>
  </div>
</Overlay>
`;
